web-client/src/main/java/no/nav/familie/webflux/builder/WebClientConfig.kt

```diff
@@ -0,0 +1,100 @@
+ */
+@Configuration
+@Import(ConsumerIdFilter::class, WebClientAutoConfiguration::class)
+class WebClientConfig {
+
+    /**
+     * Overskrever [WebClientCodecsConfiguration] fordi den ikke blir initiert av noen grunn
+     * [org.springframework.boot.autoconfigure.web.reactive.function.client.WebClientAutoConfiguration.WebClientCodecsConfiguration]
+     *
+     * Med denne blir objectMapper-bean plukket opp og brukt for encoding/decoding
+     */
+    @Bean
+    @ConditionalOnProperty("familie.web.exchangeStrategiesCustomizer", matchIfMissing = true)
+    fun exchangeStrategiesCustomizer(codecCustomizers: ObjectProvider<CodecCustomizer>): WebClientCodecCustomizer {
+        return WebClientCodecCustomizer(codecCustomizers.orderedStream().collect(Collectors.toList()))
+    }
+
+    /**
+     * Spring har default 256KB, denne settes automatisk til unlimited for å unngå trøbbel.
+     * Deaktiveres hvis man setter `spring.codec.max-in-memory-size`, til eks `100MB`
+     * Default: [org.springframework.boot.autoconfigure.http.codec.CodecsAutoConfiguration.DefaultCodecsConfiguration]
+     */
+    @Bean
+    @Order(-1)
+    @ConditionalOnProperty("spring.codec.max-in-memory-size", matchIfMissing = true, havingValue = "Umulig verdi")
+    fun codecCustomizer(): CodecCustomizer {
+        return CodecCustomizer { it.defaultCodecs().maxInMemorySize(-1) }
+    }
+
+    /**
+     * Scope: prototype - for å generere en ny webbuilder for hver gang som den blir injectad, ellers er den singleton
+     *
+     * @param connectTimeout – the max time, in milliseconds, a connection can take to connect to destinations. Zero value means infinite timeout.
+     * @param socketTimeout - the socket address resolution timeout
+     * @param requestTimeout - the max time, in milliseconds, a connection can be idle (that is, without traffic of bytes in either direction)
+     * @param webClientMetricsEnabled Pga av at MetricsWebClientCustomizer legger inn metric for hver url, og vi ikke bruker URI templates
+     *                       så fjernes MetricsWebClientCustomizer for å unngå allt for mye metrics. Kan tas i bruk med property. Hvis den hadde brukt Order hadde det vært enklere å skru den av
+     */
+    @Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    @Bean(FAMILIE_WEB_CLIENT_BUILDER)
+    fun webClientBuilder(
+        webClientBuilder: WebClient.Builder,
+        jettyResourceFactory: JettyResourceFactory,
+        consumerIdFilter: ConsumerIdFilter,
+        @Value("\${familie.web.timeout.connect:2000}") connectTimeout: Long,
+        @Value("\${familie.web.timeout.socket:15000}") socketTimeout: Long,
+        @Value("\${familie.web.timeout.requestTimeout:30000}") requestTimeout: Long,
+        @Value("\${familie.web.web-metrics.enabled:false}") webClientMetricsEnabled: Boolean): WebClient.Builder {
+        if (!ClassUtils.isPresent("org.eclipse.jetty.client.HttpClient", this::class.java.classLoader)) {
```

Comment:
Det her kjens som litt meir lågnivåkode enn vi burde trengje å ha. Treng vi denne sjekken i det heile tatt?

Reply:
Godt innspill! :) Og jeg er egentlige enig.
Årsaken til denne er at httpklientene konfigurerer timeouts annerledes, så for å ikke konfigurere både netty og jetty, så sjekkes det kun for om jetty finnes, og då konfigurerer jetty.
Jeg burde i hvert fall trekke det(lag klient) ut i en egen metode med kommentar om den delen.

https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/reactive/function/client/ClientHttpConnectorConfiguration.java

https://github.com/spring-projects/spring-framework/blob/main/spring-webflux/src/main/java/org/springframework/web/reactive/function/client/DefaultWebClientBuilder.java#L336

Reply:
Aha, da skjønner eg tanken bak 💡 Har ikkje sett stort på verken Jetty eller Netty før (utover å bruke det under panseret), men liker ikkje å stole på classloader, så fins det eit flagg, enum, tilstand eller noko vi heller kan bruke til å skilje? Eller i verste fall bruke instanceof-sjekk?